power: Rework OS <-> Application interface

pizi-nordic opened the pull request:
This PR simplifies OS <-> Application interface controlling power management. In the previous approach application-based PM required overriding sys_suspend() and sys_resume() functions. As these functions actually implemented power state change, in such case application basically had to provide own implementation of all PM-related stuff, which was not portable and hard to maintain.

This PR changes this scheme: The sys_suspend() and sys_resume() are now system functions while the application could either use built-in power management policies or provide its own. All details of power mode switching are now handled by the OS.

Also, this PR cleans up the Kconfig options related to system-level power management grouping them under common CONFIG_SYS_PM_ prefix as well as groups all power states under common naming scheme.

pizi-nordic committed:
power: Do not suspend devices if device PM is not compiled in

This commit fixes compilation of System Power Management module
when Device Power Management is disabled.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
test: kernel: device.pm: Do not enable system-level PM. It is not used.

This commit disables system-level power management in the device power
management test as it is not used.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
power: Rework OS <-> Application interface

This commit simplifies OS <-> Application interface controlling power
management. In the previous approach application-based PM required
overriding sys_suspend() and sys_resume() functions. As these functions
actually implemented power state change, in such case application
basically had to provide own implementation of all PM-related stuff,
which was not portable and hard to maintain.

This commit changes this scheme: The sys_suspend() and sys_resume()
are now system functions while the application could either use
built-in power management policies or provide its own. All details
of power mode switching are now handled by the OS.

Also, this commit cleans up the Kconfig options related to system-level
power management grouping them under common CONFIG_SYS_PM_ prefix.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
power: Clean up power state names

Some of power states used numerical suffix while otthers not.
This commit adds proper suffix to all power state names.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

f

anangl commented on subsys/power/policy/Kconfig:
```diff
@@ -1,64 +1,67 @@
-	int "LPS minimum residency"
-	depends on SYS_POWER_STATE_CPU_LPS_SUPPORTED
+config SYS_PM_LPS_0_MIN_RES
+	int "LPS_0 minimum residency"
```
What about "Low Power state X minimum residency"? And then "Deep Sleep state X ..." below.

pizi-nordic replied:
Fixed

pizi-nordic committed:
power: policy_residency: Update Kconfig options description

This commit adds better description to the Kconfig options
configuring residency-based power policy.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pdunaj commented on tests/kernel/profiling/profiling_api/prj.conf:
```diff
@@ -8,8 +8,9 @@ CONFIG_TEST_HW_STACK_PROTECTION=n
 CONFIG_SYS_POWER_LOW_POWER_STATES=y
-CONFIG_IDLE_STACK_SIZE=512
+CONFIG_SYS_PM_POLICY_APP=y
+CONFIG_IDLE_STACK_SIZE=768
```
why idle stack size is increased?

pizi-nordic replied:
The example uses hook called from idle thread to print some info. After changes the hook is called one-level deeper and there was no space for the printk().

pdunaj commented on tests/kernel/profiling/profiling_api/prj.conf:
```diff
@@ -8,8 +8,9 @@ CONFIG_TEST_HW_STACK_PROTECTION=n
 # to check isr
 CONFIG_IRQ_OFFLOAD=y
-CONFIG_ISR_STACK_SIZE=512
+CONFIG_ISR_STACK_SIZE=768
```
was increasing of isr stack needed?

pizi-nordic replied:
I followed the changes in the idle there as it looks that we were on the edge there.

mnkp reviewed (commented):
This is a partial review. As a general comment "power: Rework OS <-> Application interface" is a new feature which needs to be thoroughly reviewed and as such should probably be merged after 1.14 is released. On the other hand
- power: Clean up power state names
- power: Do not suspend devices if device PM is not compiled in
- test: kernel: device.pm: Do not enable system-level PM. It is not used.
- power: policy_residency: Update Kconfig options description

are bug fixes and should be merged for 1.14.

pizi-nordic commented on include/power.h:
```diff
@@ -27,8 +27,8 @@ enum power_states {
-# ifdef CONFIG_SYS_POWER_STATE_CPU_LPS_SUPPORTED
-	SYS_POWER_STATE_CPU_LPS,
+# ifdef CONFIG_SYS_POWER_STATE_CPU_LPS_0_SUPPORTED
+	SYS_POWER_STATE_CPU_LPS_0,
```
There is no problem to change that. We just have to decide what would be best.

carlescufi replied:
I disagree there. I think that power states starting with 0 make sense in this context, because they are mappable to indices in an array. In the Bluetooth world for example we always have packet types starting from 0, or connection types starting from 0.

mnkp replied:
In this case the numerical representation of the state will have no connection to its name. It's best to look at the definition of `enum power_states` in power.h. Having a name suggests that such connection exists is potentially dangerous.

While vendors typically provide multiple low power / deep sleep levels they never start naming them with index 0, if the indexes are used at all. That's not natural in this use case. 0 is in fact often reserved to mean an active state.

mnkp replied:
> is potentially dangerous.

I apologize for my poor English. I was writing in a hurry. There is nothing dangerous about having names starting with index 0 but it is misleading. Here are a few examples of state mapping for low power states:
- quark_se_c1000_ss: LPS_0 -> SS1, LPS_1 -> SS2
- intel_quark: LPS_0 -> C1, LPS_1 -> C2, LPS_2 -> C2LP 
- SiLabs: LPS_0 -> EM1 Sleep, LPS_1 -> EM2 DeepSleep (active state is called EM0 Active)

assuming that we proceed with the current naming convention.

Also we need to take into account that the Kernel itself puts CPU in the sleep mode by calling k_cpu_idle(). This is also a low power mode typically documented in the datasheet in the section on power management. On ARM architecture this is a mode entered by calling WFI instruction with no need for any interaction with SoC power management module. It turns off CPU clock. This is in addition to LPS_0 which is the first 'real' low power state. So we end up with one sub 0 low power state.

pizi-nordic replied:
I have renamed the states. Now they are starting from 1.
This better match the well known C0, C1, Cn naming where C0 is the active state.

pizi-nordic commented on subsys/power/policy/Kconfig:
```diff
@@ -23,45 +23,48 @@ endchoice
 	default 5
 	help
-	  Minimum residency in ticks to enter LPS state.
+	  Minimum residency in ticks to enter SYS_POWER_STATE_CPU_LPS_0 state.
```
Good catch! Thank you. Fixed.

mnkp replied:
The description was changed to
```
  Minimum residency in seconds to enter SYS_POWER_STATE_CPU_LPS_0 state.
```

I was complaining about the discrepancy. However, settling on a time unit in seconds is probably not the right choice. To exit low power state a typical MCU will need as little as a couple of microseconds. To wake up from a deep sleep state a device may take from a few hundred microseconds to a few milliseconds. The kernel interface provides a resolution of 1 tick. Having the default low power policy define the time resolution in seconds seems like too coarse a unit.

I would use here resolution in ticks or ms.

pizi-nordic replied:
Yes. I thought about that today. I think that ms is best choice.

pizi-nordic replied:
Changed to ms.

mnkp replied:
Thanks, looks good. I think we need also to update the testcases e.g. samples/boards/nrf52/power_mgr/src/main.c. We have there `k_sleep(K_SECONDS(sleep_seconds));`.

pizi-nordic replied:
I changed the example: https://github.com/zephyrproject-rtos/zephyr/pull/13317/files#diff-5f3d088ba8e6bc30050b245f06fb4afc

mnkp commented:
Is this PR meant to be merged for 1.14? I'm asking since there is not milestone attached however it looks like we are in a hurry to approve it.

nashif commented:
@mnkp Yes, this is one of the things we are seeking exception for to get into 1.14.

mnkp reviewed (commented):
It would be really nice if this PR updated also documentation, otherwise it's difficult to do a proper review. For me especially the intended purpose of device management code is unclear, e.g. usage of `core_devices`.

My other complaint is that include/power.h is mixing such different functions as `sys_pm_notify_lps_entry` which is to be used by the application and `sys_power_state_post_ops` which is part of the internal implementation of the power management subsystem. What application user sees, part of the official power API, and what we do internally should be separated.

pizi-nordic commented on subsys/power/power.c:
```diff
@@ -121,11 +122,12 @@ enum power_states sys_suspend(s32_t ticks)
 	sys_pm_debug_stop_timer();
 
+#if CONFIG_DEVICE_POWER_MANAGEMENT
 	if (deep_sleep) {
```
There was (and still is) a link between device and system power management. When you are entering DEEP_STATE mode, the system power management shutdown the devices.

Please note, that this part is far (really far) from perfect. We will be able to make it better when real device power management will be available.

mnkp replied:
A tighter integration between device and system power management definitely makes a lot of sense. Especially when looking into the future. The issue here is a different one.

When we call `sys_set_power_state()` to enter deep sleep mode we loose execution context. I.e. to wake up we will need to do system reset. The next line of code executed will be reset handler and not the `if (deep_sleep) {` statement.

pizi-nordic replied:
The sys_resume() should handle that. However I am not sure if other things are ready for such scenario. When device PM will be there I will try to use RAM retention and have deep sleep with context saving working.

mnkp replied:
> The sys_resume() should handle that.

`sys_resume()` will not be executed after waking up from the deep sleep. The next line of code executed will be in the reset handler.

> When device PM will be there I will try to use RAM retention and have deep sleep with context saving working.

Currently the description of Deep Sleep State mentions among its characteristics: "RAM is selectively retained". That's however more of a special case. The description should be updated. While there are probably SoCs that offer partial RAM retention in Deep Sleep mode that's definitely not a standard feature.

pizi-nordic replied:
The lack of proper description of power modes is a real problem.
According to code, you have 3 ways to exit deep sleep:

1) Continue execution. RAM retention is assumed.
2) Through sys_resume(): Low-level code will call this function from ISR, and then normal execution will continue. RAM retention assumed in this case.
3) Through reset. No RAM retention is assumed.

Please note, that so far I not touched low-level part of the power management. All the changes are focusing on policies and power state naming. To get into low-level stuff we need proper device power management, which is under development. When it will be done, I will get back to system-level management and I will try to make it really working.

Up to the time feel free to propose a good abstraction model of system-level power management, better description of power modes etc. There is still a lot of not-so-defined things in the area.

mnkp replied:
> According to code, you have 3 ways to exit deep sleep:
> 1. Continue execution. RAM retention is assumed.
> 2. Through sys_resume(): Low-level code will call this function from ISR, and then normal execution will continue. RAM retention assumed in this case.
> 3. Through reset. No RAM retention is assumed.

The first two options would only be required if there exists a SoC where in the Deep Sleep mode the CPU has full RAM retention but all the peripherals are powered off. I am not aware of any. But maybe I was to quick to make a judgment and we should cover also such scenario. In a typical case Deep Sleep mode means no (possibly partial) RAM retention and exit via a system reset.

There may be some confusion around vendors like Intel. On Intel Quark microcontroller there are power states called deep sleep where the state of the CPU core as well as other SoC components are preserved and exit takes a few us. That's what an ARM vendor calls a low power mode.

mnkp commented on arch/Kconfig:
```diff
@@ -357,11 +357,11 @@ config SYS_POWER_LOW_POWER_STATES_SUPPORTED
 	  This option signifies that the target supports the SYS_POWER_LOW_POWER_STATES
 	  configuration option.
```
We've changed the meaning of `SYS_POWER_LOW_POWER_STATES_SUPPORTED`. Before it indicated that a given SoC supports low power and user is free to provide an implementation of `sys_suspend` function. Now that a power management subsystem implements `sys_suspend` the `SYS_POWER_LOW_POWER_STATES_SUPPORTED` indicates that the functionality is actually implemented by the SoC. There are multiple SoCs which do not implement low power support however they do
```
...
soc/arm/st_stm32/stm32l4/Kconfig.series:	select SYS_POWER_LOW_POWER_STATES_SUPPORTED
soc/arm/st_stm32/stm32f2/Kconfig.series:	select SYS_POWER_LOW_POWER_STATES_SUPPORTED
soc/arm/atmel_sam/same70/Kconfig.series:	select SYS_POWER_LOW_POWER_STATES_SUPPORTED
...
```
These should be removed. I.e. only SoCs that provide `sys_set_power_state()` should select `SYS_POWER_LOW_POWER_STATES_SUPPORTED` or `SYS_POWER_DEEP_SLEEP_STATES_SUPPORTED`.

pizi-nordic replied:
I will check which SoC provides sys_set_power_state() interface and update Kconfigs.
Thank you for catching that.

pizi-nordic commented on subsys/power/device.c:
```diff
@@ -11,6 +11,7 @@
 #include <device.h>
 #include "policy/pm_policy.h"
 
+#if defined(CONFIG_SYS_POWER_MANAGEMENT)
```
Please look at https://github.com/zephyrproject-rtos/zephyr/pull/13317#discussion_r256772555

mnkp replied:
Thanks. However, functions defined in `subsys/power/device.c` seem to be useful also when the System Power Management is not enabled. The above `#if` forces the user to enable both System Power Management and Device Power Management to use them. That's where I miss the documentation.

pizi-nordic replied:
Have you tried to use them? They are by design bound to given hardware and not really usable for real-world scenario. So far this is placeholder where real device-power management migt be plugged in.

mnkp replied:
No, I haven't used device power management so far. 
> They are by design bound to given hardware

That's another thing I was going to ask about. Looking at `subsys/power/CMakeLists.txt` we have
```
zephyr_sources_ifdef(CONFIG_DEVICE_POWER_MANAGEMENT device.c)
```
So the `device.c` will be compiled whenever Device Power Management is enabled. However looking at the [content of the file](https://github.com/zephyrproject-rtos/zephyr/blob/4b8e602dca82a36262d61105b3c743cc507e0056/subsys/power/device.c#L36) it looks like we want to generate a compile time error if we are not compiling the code on the selected Nordic SoCs. It's true that currently also `CONFIG_SYS_POWER_MANAGEMENT` needs to be enabled for the code to be active but that still doesn't look right. There are device power management features provided by the kernel, e.g. `device_busy_set`, `device_set_power_state`. They should work just fine. And there might be projects using it. We don't offer a working replacement so we shouldn't break existing implementation. I.e. if this PR is merged as is it won't be possible to use Device Power Management without some hacks.

pizi-nordic replied:
It looks that this file introduces a lot of confusion. I will try to explain things:

1. The device power management introduces device_set_power_state() and device_get_power_state() interfaces. Nothing else.

2. Contents of the subsys/power/device.c is compiled in when **both** `CONFIG_SYS_POWER_MANAGEMENT` and `CONFIG_DEVICE_POWER_MANAGEMENT` is enabled.

3. The file **uses** Device Power Management to shut-down devices when DEEP SLEEP state is selected by System Power Management. This means that it is client of Device Power Management not used when there is no System Power Management.

mnkp replied:
> 1. The device power management introduces device_set_power_state() and device_get_power_state() interfaces. Nothing else.

Sure, but there might be users of the API which build their own device power management application. The problem with this PR is that device power management it provides is not yet production ready and at the same time...
> 2. Contents of the subsys/power/device.c is compiled in when **both** `CONFIG_SYS_POWER_MANAGEMENT` and `CONFIG_DEVICE_POWER_MANAGEMENT` is enabled.

an application which enables both options will end up with a compile time error (unless it is build for a specific Nordic SoC).
> 3. The file **uses** Device Power Management to shut-down devices when DEEP SLEEP state is selected by System Power Management. This means that it is client of Device Power Management not used when there is no System Power Management.

Looking at the functions provided by `subsys/power/device.c` it doesn't seem like the code should be exclusive to System Power Management. An application could also use services provided by Device Power Management, be its client. But surely I may be missing the big picture. I was just asking.

pizi-nordic commented on subsys/power/power.c:
```diff
@@ -101,14 +101,15 @@ enum power_states sys_suspend(s32_t ticks)
 	sys_pm_notify_lps_entry(pm_state);
 
 	if (deep_sleep) {
```
Done through IS_ENABLED() macro.

pizi-nordic commented on arch/Kconfig:
```diff
@@ -357,11 +357,11 @@ config SYS_POWER_LOW_POWER_STATES_SUPPORTED
+	  This option signifies that the target supports the SYS_POWER_STATE_CPU_LPS_0
 	  configuration option.
 
 config SYS_POWER_STATE_CPU_LPS_1_SUPPORTED
```
No problem for me, however I would like to hear others opinion in order to avoid back and forth changes.

nashif replied:
Naming needs lots of love IMO, there is:

CONFIG_SYS_PM...
CONFIG_SYS_POWER...
CONFIG_<...>_POWER_MANAGEMENT
...

same with function names and APIs.


Maybe we should just drop the sys_ altogether and using CONFIG_PM_ or CONFIG_POWER_ in a consistent way.

pizi-nordic committed:
soc: Cleanup SYS_POWER_*_SUPPORTED Kconfig options

This commit removes "select SYS_POWER_*_SUPPORTED" entries from
Kconfig on SoCs which are not implementing sys_set_power_state()
method used to control power mode.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

dbkinder reviewed (approved):
+1 for doc change, though there seem to be developer comments hoping for more...

mnkp commented:
This PR introduces major changes to the power management subsystem. Clearly a lot of effort was invested to make it happen. However, without an updated documentation this is not really beneficial to the project. In fact, it may be even disadvantageous. Are there any plans to rewrite [Power Management](https://docs.zephyrproject.org/latest/reference/power_management/index.html) section of the Zephyr API?

nashif commented:
we can work on naming and docs in a follow up PR.

mnkp commented:
> we can work on naming and docs in a follow up PR.

Still for 1.14?
